validator/client/src/main/java/tech/pegasys/teku/validator/client/ActiveKeyManager.java

```diff
@@ -214,57 +217,107 @@ DeleteKeyResult deleteExternalValidator(final Validator activeValidator) {
   public List<PostKeyResult> importValidators(
       final List<String> keystores,
       final List<String> passwords,
-      final Optional<SlashingProtectionImporter> slashingProtectionImporter) {
-    final List<PostKeyResult> importResults = new ArrayList<>();
+      final Optional<SlashingProtectionImporter> slashingProtectionImporter,
+      final Optional<DoppelgangerDetector> maybeDoppelgangerDetector) {
+    final List<Pair<Optional<BLSPublicKey>, PostKeyResult>> importResults = new ArrayList<>();
     boolean reloadRequired = false;
     for (int i = 0; i < keystores.size(); i++) {
       importResults.add(
           importValidatorFromKeystore(
               keystores.get(i), passwords.get(i), slashingProtectionImporter));
-      if (importResults.get(i).getImportStatus() == ImportStatus.IMPORTED) {
+      if (importResults.get(i).getValue().getImportStatus() == ImportStatus.IMPORTED) {
         reloadRequired = true;
       }
     }
     if (reloadRequired) {
-      validatorTimingChannel.onValidatorsAdded();
+      maybeDoppelgangerDetector
+          .map(
+              doppelgangerDetector ->
+                  doppelgangerDetector
+                      .performDoppelgangerDetection(filterImportedPubKeys(importResults))
+                      .thenAccept(
+                          doppelgangerDetected -> {
+                            if (!doppelgangerDetected.isEmpty()) {
+                              System.exit(1);
+                            }
```

Comment:
Im not sure this is what we want... if we've got a bunch of keys already loaded and i load one that is invalid, then we're going to system.exit a running validator?

Because it's batch loading, i think potentially you could justify rejecting the whole batch once the first fails, but I'm not super keen on crashing the validator...

Probably need more opinions @ajsutton

Reply:
another case that may be interesting, a user removes a key, decides they've made a mistake and re-adds to the same validator...
If slashing protection data is in place and current on import, i wonder if we could be clever...

eg. if they have slashing protection records from last epoch, there's not much point checking if they were active previous epoch, because we cant sign anyway... not a fully formed thought...

Reply:
hmm, yeah probably not good to exit in that case. It's hard to know how to make it obvious that the key was disabled because of doppelgänger detection though - a single log line would be easy to miss. Might be worth seeing what other clients do in this case.


Not sure if we should trust slashing protection records though. You may have removed the validator then added it to a different node and then also try to readd it to this one by mistake. I think in general if you enable doppelgänger support, you're saying that you want to skip 2 epochs after a new key is added in order to be maximally safe.

Reply:
I'd happily reject the batch, that'd be in line with other import failures...

Reply:
hmm, yeah possibly, but most of the time when we reject a batch we also return an API error response but in this case we're already well past that point. So it's really a question of whether it makes it more obvious there's a problem if none of the validators activate or if that just makes it cost more...

Reply:
@rolfyone regarding the keys rejection, as mentioned by @ajsutton: The API response doesn't wait for the doppelganger check result (Which could take up to 2 epochs and hence makes it unreasonable). It returns the result of the keys imports only in the current case and the doppelganger check is run separately.
One option here, instead of shutting down the whole VC, could be to reject the detected doppelganger keys or the whole batch containing it and that would be reported in the logs only (And not in the API response).
I think we can assume that when a user activates the doppelganger detection, he should be aware of this behaviour and hence check the logs for further infos.

Reply:
@rolfyone regarding the slashing protection data, I am not sure how we could use it for this purpose. The idea here is to check if the keys are used elsewhere in the network and not locally. Otherwise we could reject any keys that have been used even after a restart. Let me know if I am missing something here

Reply:
Given Lighthouse is doing a full exit in this case and on the basis that users activated doppelgänger because they wanted to prioritise safety over availability I think I've come around to exiting.

I'd suggest not calling `System.exit` directly in the code though - it's better to inject an instance of an object that calls exit so that we can replace it with a mock in tests and check it was told to exit without actually exiting the test JVM.

Reply:
Ok we can probably document the behaviour I guess. It seems like pretty bad UX, If you're using keymanager you and dopelganger you'll be restricted from any rewards for a further 2 epochs, and none of the validators will be active until you resolve your issue...